File: ios-app/Plantgotchi/Models/StrainProfile.swift
import Foundation
import GRDB

/// A cannabis strain with growing metadata.
struct StrainProfile: Codable, Identifiable, Equatable {
    var id: String
    var name: String
    var type: StrainType?
    var flowerWeeksMin: Int?
    var flowerWeeksMax: Int?
    var difficulty: String?
    var thresholdsByPhase: String?
    var notes: String?
    var isCustom: Bool
    var userId: String?
    var createdAt: String?

    init(
        id: String = UUID().uuidString,
        name: String,
        type: StrainType? = nil,
        flowerWeeksMin: Int? = nil,
        flowerWeeksMax: Int? = nil,
        difficulty: String? = nil,
        thresholdsByPhase: String? = nil,
        notes: String? = nil,
        isCustom: Bool = false,
        userId: String? = nil,
        createdAt: String? = nil
    ) {
        self.id = id
        self.name = name
        self.type = type
        self.flowerWeeksMin = flowerWeeksMin
        self.flowerWeeksMax = flowerWeeksMax
        self.difficulty = difficulty
        self.thresholdsByPhase = thresholdsByPhase
        self.notes = notes
        self.isCustom = isCustom
        self.userId = userId
        self.createdAt = createdAt
    }
}

// MARK: - GRDB Records

extension StrainProfile: TableRecord, FetchableRecord, PersistableRecord {
    static let databaseTableName = "strain_profiles"

    enum Columns: String, ColumnExpression {
        case id
        case name
        case type
        case flowerWeeksMin = "flower_weeks_min"
        case flowerWeeksMax = "flower_weeks_max"
        case difficulty
        case thresholdsByPhase = "thresholds_by_phase"
        case notes
        case isCustom = "is_custom"
        case userId = "user_id"
        case createdAt = "created_at"
    }

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case type
        case flowerWeeksMin = "flower_weeks_min"
        case flowerWeeksMax = "flower_weeks_max"
        case difficulty
        case thresholdsByPhase = "thresholds_by_phase"
        case notes
        case isCustom = "is_custom"
        case userId = "user_id"
        case createdAt = "created_at"
    }
}

File: android-app/app/src/main/java/com/plantgotchi/app/model/StrainProfile.kt
package com.plantgotchi.app.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "strain_profiles",
    indices = [
        Index(value = ["name"]),
        Index(value = ["user_id"])
    ]
)
data class StrainProfile(
    @PrimaryKey
    val id: String,

    val name: String,

    val type: String? = null, // indica, sativa, hybrid

    @ColumnInfo(name = "flower_weeks_min")
    val flowerWeeksMin: Int? = null,

    @ColumnInfo(name = "flower_weeks_max")
    val flowerWeeksMax: Int? = null,

    val difficulty: String? = null,

    @ColumnInfo(name = "thresholds_by_phase")
    val thresholdsByPhase: String? = null, // JSON string

    val notes: String? = null,

    @ColumnInfo(name = "is_custom")
    val isCustom: Boolean = false,

    @ColumnInfo(name = "user_id")
    val userId: String? = null,

    @ColumnInfo(name = "created_at")
    val createdAt: String? = null,
)
